### PR TITLE
Read feed correctly handle 410 gone.

### PR DESCRIPTION
**Pull request by j82w**
# Pull Request Template

## Description
This fixes the issue where the collection recreate causes a 410 exception to be returned to users.

InvalidPartitionExceptionRetryPolicy refreshes the name collection cache, but no the RID based Collection Cache. The PartitionKeyRangeHandler does a collection cache lookup based on rid which always gets the stale value. Updated PartitionKeyRangeHandler to use name cache lookup to match the other retry policies.

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/Handler/PartitionKeyRangeHandler.cs [hunk 1]
```diff
@@ -83,8 +83,10 @@ public PartitionKeyRangeHandler(CosmosClient client, PartitionRoutingHelper part
 
                 PartitionKeyRangeCache routingMapProvider = await this.client.DocumentClient.GetPartitionKeyRangeCacheAsync();
                 CollectionCache collectionCache = await this.client.DocumentClient.GetCollectionCacheAsync();
-                CosmosContainerSettings collectionFromCache =
-                    await collectionCache.ResolveCollectionAsync(serviceRequest, CancellationToken.None);
+                CosmosContainerSettings collectionFromCache = await collectionCache.ResolveByNameAsync(
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/Handler/PartitionKeyRangeHandler.cs]: ResolveByNameAsync doesnt take into account that the ResourceAddress could be RID based and doesn't check for `request.ForceNameCacheRefresh`, and also, the `request.Headers[HttpConstants.HttpHeaders.Version]` is not being passed, like ResolveByNameAsync does internally. 
My question is, this change reduces the scenarios where the collection can be resolved, are these reduced scenarios enough to solve all the cases that can come through here?

Comment (kirankumarkolli) [reply Microsoft.Azure.Cosmos/src/Handler/PartitionKeyRangeHandler.cs]: NameCacheRetryHandleris reseting the stale context on Request would be right thing to do.

---
In reply to: [291232719](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/377#discussion_r291232719) [](ancestors = 291232719)

File: Microsoft.Azure.Cosmos/src/Handler/CosmosRequestMessage.cs [hunk 2]
```diff
@@ -85,7 +85,7 @@ public virtual Stream Content
 
         internal DocumentServiceRequest DocumentServiceRequest { get; set; }
 
-        internal IDocumentClientRetryPolicy DocumentClientRetryPolicy { get; set; }
+        internal Action<DocumentServiceRequest> OnBeforeRequestHandler { get; set; }
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Handler/CosmosRequestMessage.cs]: Nicely done.

File: Microsoft.Azure.Cosmos/src/Routing/InvalidPartitionExceptionRetryPolicy.cs [hunk 3]
```diff
@@ -87,7 +80,12 @@ internal class InvalidPartitionExceptionRetryPolicy : IDocumentClientRetryPolicy
                 {
                     if (!string.IsNullOrEmpty(resourceIdOrFullName))
                     {
-                        this.clientCollectionCache.Refresh(resourceIdOrFullName);
+                        if(this.request == null)
+                        {
+                            throw new InvalidOperationException("OnBeforeSendRequest was not called before retry.");
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/Routing/InvalidPartitionExceptionRetryPolicy.cs]: not: The other RetryPolicies do a Trace, not a throw, not sure if we want to be consitent or not

Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/Routing/InvalidPartitionExceptionRetryPolicy.cs]: Or a Debug.Assert

Comment (j82w) [reply Microsoft.Azure.Cosmos/src/Routing/InvalidPartitionExceptionRetryPolicy.cs]: This retry policy requires the request to do the retry. Doing a trace will just cause us to miss the issue where is it not being called, and will causes users to see unexpected errors that will be very difficult to figure out.

Comment (kirankumarkolli) [reply Microsoft.Azure.Cosmos/src/Routing/InvalidPartitionExceptionRetryPolicy.cs]: In that case at-least Debug.Assert is must.

File: Microsoft.Azure.Cosmos/src/Handler/PartitionKeyRangeHandler.cs (hunk 1 above)
Comment (j82w) [reply Microsoft.Azure.Cosmos/src/Handler/PartitionKeyRangeHandler.cs]: Updated the latest iteration to just have NameCacheRetryHandleris  to force the cache refresh.

**Pull request merged**